kythe/cxx/verifier/souffle_interpreter.cc

```diff
@@ -0,0 +1,163 @@
+  }
+
+  const std::string& getName() const override {
+    static const std::string name = "kythe";
+    return name;
+  }
+};
+
+class KytheWriteStream : public souffle::WriteStream {
+ public:
+  KytheWriteStream(const std::map<std::string, std::string>& rw_operation,
+                   const souffle::SymbolTable& symbol_table,
+                   const souffle::RecordTable& record_table,
+                   std::vector<std::pair<int, int>>* output)
+      : souffle::WriteStream(rw_operation, symbol_table, record_table),
+        output_(output) {}
+
+ protected:
+  void writeNullary() override {}
+
+  void writeNextTuple(const souffle::RamDomain* tuple) override {
+    output_->push_back({tuple[0], tuple[1]});
+  }
+
+ private:
+  std::vector<std::pair<int, int>>* output_;
+};
+
+class KytheWriteStreamFactory : public souffle::WriteStreamFactory {
+ public:
+  souffle::Own<souffle::WriteStream> getWriter(
+      const std::map<std::string, std::string>& rw_operation,
+      const souffle::SymbolTable& symbol_table,
+      const souffle::RecordTable& record_table) override {
+    auto output = rw_operation.find("id");
+    CHECK(output != rw_operation.end());
+    size_t output_id;
+    CHECK(absl::SimpleAtoi(output->second, &output_id));
+    CHECK(output_id < outputs_.size());
+    return souffle::mk<KytheWriteStream>(rw_operation, symbol_table,
+                                         record_table, &outputs_[output_id]);
+  }
+
+  const std::string& getName() const override {
+    static const std::string name = "kythe";
+    return name;
+  }
+
+  ~KytheWriteStreamFactory() override = default;
+
+  size_t NewOutput() {
+    outputs_.push_back({});
+    return outputs_.size() - 1;
+  }
+
+  const std::vector<std::pair<int, int>>& GetOutput(size_t id) {
+    return outputs_[id];
+  }
+
+ private:
+  std::vector<std::vector<std::pair<int, int>>> outputs_;
+};
+// TODO(zarko): This is a temporary hack that only demonstrates that the
+// Souffle library is working. It bakes in the inputs and outputs for a
+// simple example program and checks that the interpreter calculates the
+// correct result. `Kythe{Write,Read}StreamFactory` will need to be moved
+// to separate files and updated to accept the Datalog relations that the
+// verifier -> datalog compiler produces.
+bool RunSouffle() {
+  auto write_stream_factory = std::make_shared<KytheWriteStreamFactory>();
+  size_t output_id = write_stream_factory->NewOutput();
+  souffle::IOSystem::getInstance().registerWriteStreamFactory(
+      write_stream_factory);
+  souffle::IOSystem::getInstance().registerReadStreamFactory(
+      std::make_shared<KytheReadStreamFactory>());
+  auto ram_tu =
+      souffle::ParseTransform(absl::StrCat(R"(
+        .decl edge(x:number, y:number)
+        .input edge(IO=kythe)
+
+        .decl path(x:number, y:number)
+        .output path(IO=kythe, id=)",
+                                           std::to_string(output_id), R"()
```

Comment:
nit: prefer StrCat over std::to_string (the former covers all of the uses of the latter while generally being faster)